appstudio-controller/controllers/appstudio.redhat.com/environment_controller.go

```diff
@@ -39,21 +45,146 @@ type EnvironmentReconciler struct {
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the Environment object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
 func (r *EnvironmentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	_ = log.FromContext(ctx)
 
-	fmt.Println("Environment event: ", req)
+	// The goal of this function is to ensure that if an Environment exists, and that Environment
+	// has the 'kubernetesCredentials' field defined, that a corresponding
+	// GitOpsDeploymentManagedEnvironment exists (and is up-to-date).
+	environment := &appstudioshared.Environment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(environment), environment); err != nil {
+
+		if apierr.IsNotFound(err) {
+			// A) The Environment resource could not be found: the owner reference on the GitOpsDeploymentManagedEnvironment
+			// should ensure that it is cleaned up, so no more work is required.
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, fmt.Errorf("unable to retrieve Environment: %v", err)
+	}
+
+	desiredManagedEnv, err := generateDesiredResource(ctx, *environment, r.Client)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to generate expected GitOpsDeploymentManagedEnvironment resource: %v", err)
+	}
+	if desiredManagedEnv == nil {
+		return ctrl.Result{}, nil
+	}
+
+	currentManagedEnv := generateEmptyManagedEnvironment(environment.Name, environment.Namespace)
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(&currentManagedEnv), &currentManagedEnv); err != nil {
+
+		if apierr.IsNotFound(err) {
+			// B) The GitOpsDeploymentManagedEnvironment doesn't exist, so needs to be created.
+
+			if err := r.Client.Create(ctx, desiredManagedEnv); err != nil {
+				return ctrl.Result{}, fmt.Errorf("unable to create new GitOpsDeploymentManagedEnvironment: %v", err)
+			}
+
+			// Success: the resource has been created.
```

Comment:
Same question here, do we need a `log.Info` informing that the object is created?